src/backend/mpi/swdsm.cpp

```diff
@@ -507,37 +540,34 @@ void handler(int sig, siginfo_t *si, void *context){
 	std::size_t startIndex = getCacheIndex(aligned_access_offset);
 
 	/* Get homenode and offset, protect with ibsem if first touch */
-	argo::node_id_t homenode;
-	std::size_t offset;
-	if(dd::is_first_touch_policy()){
-		std::lock_guard<std::mutex> lock(spin_mutex);
-		sem_wait(&ibsem);
-		homenode = get_homenode(aligned_access_offset);
-		offset = get_offset(aligned_access_offset);
-		sem_post(&ibsem);
-	}else{
-		homenode = get_homenode(aligned_access_offset);
-		offset = get_offset(aligned_access_offset);
-	}
+	argo::node_id_t homenode = get_homenode(aligned_access_offset);
+	std::size_t offset = get_offset(aligned_access_offset);
 
 	std::uint64_t id = static_cast<std::uint64_t>(1) << getID();
 	std::uint64_t invid = ~id;
-
-	pthread_mutex_lock(&cachemutex);
+	std::size_t sharer_win_offset = get_sharer_win_offset(classidx);
+
+	/* Acquire shared sync lock and first cache index lock */
+	double sync_lock_start = MPI_Wtime();
+	pthread_rwlock_rdlock(&sync_lock);
+	double sync_lock_end = MPI_Wtime();
+	{
```

Comment:
What is the purpose of the code block here?

Reply:
the lock_guard needs to destruct to release the lock